reframe/core/buildsystems.py

```diff
@@ -779,6 +782,89 @@ def generated_modules(self):
         return self._eb_modules
 
 
+class Spack(BuildSystem):
+    packages = fields.TypedField(typ.List[str])
+    environment = fields.TypedField(str, type(None))
+    upstreams = fields.TypedField(typ.List[str])
+
+    def __init__(self):
+        super().__init__()
+        self.packages = []
+        self.environment = None
+        self.upstreams = []
+        self._prefix_save = None
+
+    def _prepare_scope(self):
+        os.mkdir('rfm_scope')
+        scope_config = {
+            'config': {
+                'install_tree': 'spack/opt/spack',
+                'module_roots': {
+                    'tcl': 'spack/share/spack/modules',
+                    'lmod': 'spack/share/spack/lmod'
+                }
+            }
+        }
+        with open('rfm_scope/config.yaml', 'w') as fp:
+            yaml.dump(scope_config, fp, default_flow_style=False)
+
+        scope_upstreams = {'upstreams': {}}
+        for i, upstream in enumerate(self.upstreams, 1):
+            scope_upstreams['upstreams'].update({
+                f'spack-instance-{i}': {'install_tree': upstream}
+            })
+
+        with open('rfm_scope/upstreams.yaml', 'w') as fp:
+            yaml.dump(scope_upstreams, fp, default_flow_style=False)
+
+    def _emit_build_commands_pkgs(self, environ):
+        self._prepare_scope()
+        return [f'spack -C rfm_scope install {pkg}'
+                for pkg in self.packages]
+
+    def _emit_build_commands_env(self, environ):
+        return[f'spack env activate -d {self.environment}',
+               'spack install']
+
+    def emit_build_commands(self, environ):
+        self._prefix_save = os.getcwd()
+        if self.packages:
+            return self._emit_build_commands_pkgs(environ)
+        else:
+            return self._emit_build_commands_env(environ)
+
+    def post_build(self, buildjob):
+        # Store the modules generated by Spack
+
+        with open(os.path.join(self._prefix_save, buildjob.stdout)) as fp:
+            out = fp.read()
+
+        matches = re.findall(r'\[\+\] \S+/spack/opt/spack/(\S+)/\S+/(\S+)',
+                             out)
+
+        self._spack_modules = [{'name': os.path.basename(mod),
+                                'collection': False,
+                                'path': None}
+                               for _, mod in matches]
+
+        spack_arch = matches[0][0]
```

Comment:
This `spack_arch` is not really the architecture used in your build because it you are using Lmod, then it will be always `x86_64`, so I think it would be better to name it `spack_module_arch` or just `module_arch`.
Furthermore there is an additional `Core` folder for Lmod, if one is using the hierarchical naming scheme...